source/spawn/spawn-channels.ads

```diff
@@ -0,0 +1,101 @@
+------------------------------------------------------------------------------
+--                         Language Server Protocol                         --
+--                                                                          --
+--                     Copyright (C) 2018-2022, AdaCore                     --
+--                                                                          --
+-- This is free software;  you can redistribute it  and/or modify it  under --
+-- terms of the  GNU General Public License as published  by the Free Soft- --
+-- ware  Foundation;  either version 3,  or (at your option) any later ver- --
+-- sion.  This software is distributed in the hope  that it will be useful, --
+-- but WITHOUT ANY WARRANTY;  without even the implied warranty of MERCHAN- --
+-- TABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public --
+-- License for  more details.  You should have  received  a copy of the GNU --
+-- General  Public  License  distributed  with  this  software;   see  file --
+-- COPYING3.  If not, go to http://www.gnu.org/licenses for a complete copy --
+-- of the license.                                                          --
+--                                                                          --
+-- As a special exception under Section 7 of GPL version 3, you are granted --
+-- additional permissions described in the GCC Runtime Library Exception,   --
+-- version 3.1, as published by the Free Software Foundation.               --
+------------------------------------------------------------------------------
+--  It is Glib's version of the package.
+
+with Ada.Streams;
+
+with Glib.IOChannel;
+private with Glib.Main;
+
+limited with Spawn.Internal;
+
+private package Spawn.Channels is
+
+   type Channels (Process : not null access Spawn.Internal.Process'Class) is
+     private;
+
+   type IO_Event_Callback is
+     access procedure (Self : in out Spawn.Internal.Process'Class);
+
+   procedure Setup_Channels
+     (Self                : in out Channels;
+      Standard_Input_PTY  : Boolean;
+      Standard_Output_PTY : Boolean);
```

Comment:
Perhaps it worth to have a comment here and mention that stderr is always a pip. I would mention also that stdin/stdout share the same IOChannel/FD and this is the reason why we have a single channel set object instead of three separate channel objects.

Reply:
Stderr can be PTY too, this was added by one of following patch.